File: backend/dockerfiles/execute.cpp
#!/bin/bash
# C++ code execution script for GhostIDE
# Executes user code with security restrictions

set -e

# Read code from stdin
CODE=$(cat)

# Create temporary directory
TEMP_DIR=$(mktemp -d)
cd "$TEMP_DIR"

# Write code to main.cpp
echo "$CODE" > main.cpp

# Compile the code
if ! g++ -o main main.cpp -std=c++17 2>&1; then
    echo "Compilation failed" >&2
    rm -rf "$TEMP_DIR"
    exit 1
fi

# Execute with timeout
timeout 30s ./main 2>&1 || {
    exit_code=$?
    if [ $exit_code -eq 124 ]; then
        echo "Error: Code execution timed out (30 seconds)" >&2
        exit 1
    else
        exit $exit_code
    fi
}

# Clean up
rm -rf "$TEMP_DIR"

File: backend/dockerfiles/execute.py
#!/usr/bin/env python3
"""
Python code execution script for GhostIDE
Executes user code with security restrictions
"""

import sys
import os
import subprocess
import tempfile
import signal
from pathlib import Path

def execute_code():
    """Execute Python code from stdin"""
    try:
        # Read code from stdin
        code = sys.stdin.read()
        
        # Create temporary file for code
        with tempfile.NamedTemporaryFile(mode='w', suffix='.py', delete=False) as f:
            f.write(code)
            temp_file = f.name
        
        try:
            # Execute the code with timeout
            result = subprocess.run(
                [sys.executable, temp_file],
                capture_output=True,
                text=True,
                timeout=30,  # 30 second timeout
                cwd='/tmp'  # Run in safe directory
            )
            
            # Output results
            if result.stdout:
                print(result.stdout, end='')
            if result.stderr:
                print(result.stderr, end='', file=sys.stderr)
            
            sys.exit(result.returncode)
            
        finally:
            # Clean up temporary file
            os.unlink(temp_file)
            
    except subprocess.TimeoutExpired:
        print("Error: Code execution timed out (30 seconds)", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {str(e)}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    execute_code()